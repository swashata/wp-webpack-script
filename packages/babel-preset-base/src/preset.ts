// tslint:disable no-any

export interface PresetOptions {
	noDynamicImport?: boolean;
	noImportMeta?: boolean;
	noClassProperties?: boolean;
	noJsonStrings?: boolean;
	noRuntime?: boolean;
	hasReact?: boolean;
	useReactJsxRuntime?: boolean;
	presetEnv?: {};
	presetReact?: {};
	[x: string]: any;
}

export interface BabelPluginConfig {
	[x: string]: any;
}

export type singleBabelPlugin = string | [string, BabelPluginConfig];

export interface PossiblePlugins {
	[x: string]: singleBabelPlugin;
}

export type babelPresetConfiguration = {
	[x: string]: boolean | string | babelPresetConfiguration;
};
export type babelPreset = [string] | [string, babelPresetConfiguration];

export const preset = (opts: PresetOptions | null = {}) => {
	// from CRA
	// This is similar to how `env` works in Babel:
	// https://babeljs.io/docs/usage/babelrc/#env-option
	// We are not using `env` because it’s ignored in versions > babel-core@6.10.4:
	// https://github.com/babel/babel/issues/4539
	// https://github.com/facebook/create-react-app/issues/720
	// It’s also nice that we can enforce `NODE_ENV` being specified.
	const env = process.env.BABEL_ENV || process.env.NODE_ENV;
	// Extract this preset specific options and pass the rest to @babel/preset-env
	const {
		presetEnv = {},
		presetReact = {},
		hasReact = true,
		useReactJsxRuntime = false,
		// Put everything else inside noPlugins, which we will use later
		// to put or cancel out stage-3 plugins.
		// noDynamicImport = false,
		// noImportMeta = false,
		// noClassProperties = false,
		// noJsonStrings = false,
		...noPlugins
	} = opts || {};

	// Create the presets
	const presets: babelPreset[] = [
		[
			require.resolve('@babel/preset-env'),
			{ modules: env === 'test' ? 'commonjs' : false, ...presetEnv },
		],
	];
	// Add react if needed
	if (hasReact) {
		presets.push([
			require.resolve('@babel/preset-react'),
			{
				// Put development based on BABEL_ENV
				// Adds component stack to warning messages
				// Adds __self attribute to JSX which React will use for some warnings
				development: env !== 'production',
				// Will use the native built-in instead of trying to polyfill
				// behavior for any plugins that require one.
				useBuiltIns: true,
				runtime: useReactJsxRuntime ? 'automatic' : 'classic',
				// But spread later, so that user can override it
				...presetReact,
			},
		]);
	}

	// Create the plugins
	const plugins: singleBabelPlugin[] = [];
	const wannabePlugins: PossiblePlugins = {
		noDynamicImport: require.resolve('@babel/plugin-syntax-dynamic-import'),
		noImportMeta: require.resolve('@babel/plugin-syntax-import-meta'),
		noClassProperties: [
			require.resolve('@babel/plugin-proposal-class-properties'),
			{ loose: false },
		],
		noJsonStrings: require.resolve('@babel/plugin-proposal-json-strings'),
		noRuntime: [
			require.resolve('@babel/plugin-transform-runtime'),
			{
				corejs: false,
				helpers: true,
				regenerator: true,
				// We might wanna turn it on once node LTS has ESModules support
				useESModules: env !== 'test',
			},
		],
		noOptionalChaining: require.resolve(
			'@babel/plugin-proposal-optional-chaining'
		),
		noNullishCoalescingOperator: require.resolve(
			'@babel/plugin-proposal-nullish-coalescing-operator'
		),
	};
	// Add them, only if user hasn't explicitly disabled it
	Object.keys(wannabePlugins).forEach((pKey: string) => {
		if (noPlugins[pKey] !== true) {
			plugins.push(wannabePlugins[pKey]);
		}
	});

	// Necessary to include regardless of the environment because
	// in practice some other transforms (such as object-rest-spread)
	// don't work without it: https://github.com/babel/babel/issues/7215
	plugins.push([
		require.resolve('@babel/plugin-transform-destructuring'),
		{
			// Use loose mode for performance:
			// https://github.com/facebook/create-react-app/issues/5602
			loose: false,
			selectiveLoose: [
				'useState',
				'useEffect',
				'useContext',
				'useReducer',
				'useCallback',
				'useMemo',
				'useRef',
				'useImperativeHandle',
				'useLayoutEffect',
				'useDebugValue',
			],
		},
	]);

	// Return the preset and some of stage-3 plugins
	// We will remove them, once it becomes stage-4, i.e included in preset-env
	return {
		presets,
		plugins,
	};
};
