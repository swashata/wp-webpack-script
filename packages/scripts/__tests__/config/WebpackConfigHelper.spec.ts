import { PresetOptions } from '@wpackio/babel-preset-base/lib/preset';
// tslint:disable-next-line:no-implicit-dependencies
import ForkTsCheckerWebpackPlugin from 'fork-ts-checker-webpack-plugin';
import miniCssExtractPlugin from 'mini-css-extract-plugin';
import path from 'path';
import webpack from 'webpack';
import { webpackOptionsOverrideFunction } from '../../src/config/project.config.default';
import { WebpackConfigHelper } from '../../src/config/WebpackConfigHelper';
import { initConfig, projectConfig, serverConfig } from '../helpers/testConfig';
import {
	findWpackIoBabelOnJs,
	findWpackIoBabelOnTJs,
	findWpackIoBabelOnTs,
	getConfigFromProjectAndServer,
} from '../helpers/testUtils';

beforeEach(initConfig);

describe('CreateWebPackConfig', () => {
	// Now do the testing

	test('can be instantiated with proper config', () => {
		const cwc = new WebpackConfigHelper(
			projectConfig.files[0],
			getConfigFromProjectAndServer(projectConfig, serverConfig),
			'/foo/bar',
			true
		);
		expect(cwc).toBeInstanceOf(WebpackConfigHelper);
	});

	// cwc.getEntry()
	describe('getEntry', () => {
		test('has webpack hot client with file `name` when isDev == true', () => {
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				getConfigFromProjectAndServer(projectConfig, serverConfig),
				'/foo/bar',
				true
			);
			const entry = cwc.getEntry();
			Object.keys(entry).forEach(key => {
				expect(Array.isArray(entry[key])).toBeTruthy();
				const hotClient = entry[key].pop();
				expect(hotClient).toMatch(/^webpack\-hot\-middleware/);
				expect(hotClient).toMatch(/name\=config1/);
			});
		});
		test('does not have webpack hot client but publicPath wntrypoint when isDev != true', () => {
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				getConfigFromProjectAndServer(projectConfig, serverConfig),
				'/foo/bar',
				false
			);
			const entry = cwc.getEntry();
			Object.keys(entry).forEach(key => {
				expect(Array.isArray(entry[key])).toBeTruthy();
				expect(entry[key].pop()).not.toMatch(
					/^webpack\-hot\-middleware/
				);
				expect(entry[key][0]).toBe(`@wpackio/entrypoint/lib/index`);
			});
		});
	});

	// cwc.getOutput()
	describe('getOutput', () => {
		test('considers file `name` for outputpath', () => {
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				getConfigFromProjectAndServer(projectConfig, serverConfig),
				'/foo/bar',
				false
			);
			const output = cwc.getOutput();
			expect(output.path).toMatch(/\/dist$/);
			expect(output.filename).toBe('config1/[name].js');
		});
		test('has empty publicPath on production build', () => {
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				getConfigFromProjectAndServer(projectConfig, serverConfig),
				'/foo/bar',
				false
			);
			const output = cwc.getOutput();
			expect(output.publicPath).toBe('');
		});
		test('has proper publicPath for developmenr server', () => {
			const devCwc = new WebpackConfigHelper(
				projectConfig.files[0],
				getConfigFromProjectAndServer(projectConfig, serverConfig),
				'/foo/bar',
				true
			);
			const devOutput = devCwc.getOutput();
			expect(devOutput.publicPath).toMatch(/^\/\/(.*)\/dist\/$/);
		});
		test('respects type for constructing publicPath on dev server', () => {
			const devCwc = new WebpackConfigHelper(
				projectConfig.files[0],
				getConfigFromProjectAndServer(projectConfig, serverConfig),
				'/foo/bar',
				true
			);
			const devOutput = devCwc.getOutput();
			expect(devOutput.publicPath).toMatch(
				/^\/\/(.*)\/wp-content\/plugins\/(.*)\/dist\/$/
			);
		});
	});

	// getPlugins()
	describe('getPlugins', () => {
		test('has proper plugins for build mode', () => {
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				getConfigFromProjectAndServer(projectConfig, serverConfig),
				'/foo/bar',
				false
			);
			const plugins = cwc.getPlugins();
			expect(plugins).toMatchSnapshot();
		});
		test('has proper plugins for dev mode', () => {
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				getConfigFromProjectAndServer(projectConfig, serverConfig),
				'/foo/bar',
				true
			);
			const plugins = cwc.getPlugins();
			expect(plugins).toMatchSnapshot();
		});
		test('has fork ts checker when tsconfig.json is present', () => {
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				getConfigFromProjectAndServer(projectConfig, serverConfig),
				path.resolve(__dirname, '../../'), // it's a hack cause the project has tsconfig.json
				true
			);
			const plugins = cwc.getPlugins();
			let hasTsChecker = false;
			plugins.forEach(plug => {
				if (plug instanceof ForkTsCheckerWebpackPlugin) {
					hasTsChecker = true;
				}
			});
			expect(hasTsChecker).toBeTruthy();
		});
	});

	// getModule()
	describe('getModule', () => {
		describe('babel-loader for typescript and javascript', () => {
			test('has babel-loader', () => {
				const cwc = new WebpackConfigHelper(
					projectConfig.files[0],
					getConfigFromProjectAndServer(projectConfig, serverConfig),
					'/foo/bar',
					true
				);
				const modules = cwc.getModule();
				if (Array.isArray(modules.rules)) {
					const jsTsRules = findWpackIoBabelOnTJs(modules);
					expect(jsTsRules).toHaveLength(2);
					jsTsRules.forEach(rule => {
						if (rule && rule.use) {
							expect(rule.use[0].loader).toBe('babel-loader');
						} else {
							throw new Error('JavaScript rule is undefined');
						}
					});
				} else {
					throw new Error('Module is not an array');
				}
			});
			test('obeys hasFlow & hasRect', () => {
				const cwc = new WebpackConfigHelper(
					projectConfig.files[0],
					{
						...getConfigFromProjectAndServer(
							projectConfig,
							serverConfig
						),
						hasFlow: true,
						hasReact: true,
					},
					'/foo/bar',
					true
				);
				const modules = cwc.getModule();
				if (Array.isArray(modules.rules)) {
					const jsRule = findWpackIoBabelOnJs(modules);
					const tsRule = findWpackIoBabelOnTs(modules);
					expect(jsRule).toHaveLength(1);
					expect(tsRule).toHaveLength(1);
					[...jsRule, ...tsRule].forEach(rule => {
						if (rule && rule.use && rule.use[0].options) {
							expect(
								rule.use[0].options.presets[0][1]
							).toMatchObject({ hasReact: true });
						} else {
							throw new Error('babel rule is undefined');
						}
					});
					if (jsRule[0].use[0].options) {
						expect(jsRule[0].use[0].options.presets[1]).toEqual([
							'@babel/preset-flow',
						]);
					} else {
						throw new Error(
							'JavaScript babel-loader options not present'
						);
					}
				} else {
					throw new Error('Module is not an array');
				}
			});

			test('overrides @wpackio/babel-preset-base from config', () => {
				const override: PresetOptions = {
					noImportMeta: true,
					presetEnv: { modules: 'umd' },
					presetReact: { pragma: 'wp.createElement' },
				};
				const cwc = new WebpackConfigHelper(
					projectConfig.files[0],
					{
						...getConfigFromProjectAndServer(
							projectConfig,
							serverConfig
						),
						jsBabelPresetOptions: override,
						tsBabelPresetOptions: override,
					},
					'/foo/bar',
					true
				);

				const modules = cwc.getModule();
				if (Array.isArray(modules.rules)) {
					const jsTsRules = findWpackIoBabelOnTJs(modules);
					expect(jsTsRules).toHaveLength(2);
					jsTsRules.forEach(rule => {
						if (rule && rule.use && rule.use[0].options) {
							expect(
								rule.use[0].options.presets[0][1]
							).toMatchObject(override);
						} else {
							throw new Error('JavaScript rule is undefined');
						}
					});
				} else {
					throw new Error('Module is not an array');
				}
			});

			test('overrides all babel-loader options from config object', () => {
				const override: webpack.RuleSetLoader['options'] = {
					presets: 'foo',
					plugins: ['bar', 'baz'],
				};
				const cwc = new WebpackConfigHelper(
					projectConfig.files[0],
					{
						...getConfigFromProjectAndServer(
							projectConfig,
							serverConfig
						),
						jsBabelOverride: override,
						tsBabelOverride: override,
					},
					'/foo/bar',
					true
				);
				const modules = cwc.getModule();
				if (Array.isArray(modules.rules)) {
					const jsTsRules = findWpackIoBabelOnTJs(modules);
					expect(jsTsRules).toHaveLength(2);
					jsTsRules.forEach(rule => {
						if (rule && rule.use && rule.use[0].options) {
							expect(rule.use[0].options).toMatchObject(override);
						} else {
							throw new Error('JavaScript rule is undefined');
						}
					});
				} else {
					throw new Error('Module is not an array');
				}
			});

			test('overrides all babel-loader options from config function', () => {
				const override: webpackOptionsOverrideFunction = defaults => {
					if (typeof defaults === 'string') {
						return defaults;
					}
					return {
						...defaults,
						plugins: ['react-hot-loader/babel'],
					};
				};
				const cwc = new WebpackConfigHelper(
					projectConfig.files[0],
					{
						...getConfigFromProjectAndServer(
							projectConfig,
							serverConfig
						),
						jsBabelOverride: override,
						tsBabelOverride: override,
					},
					'/foo/bar',
					true
				);
				const modules = cwc.getModule();
				if (Array.isArray(modules.rules)) {
					const jsTsRules = findWpackIoBabelOnTJs(modules);
					expect(jsTsRules).toHaveLength(2);
					jsTsRules.forEach(rule => {
						if (rule && rule.use && rule.use[0].options) {
							expect(rule.use[0].options).toMatchObject({
								plugins: ['react-hot-loader/babel'],
							});
						} else {
							throw new Error('JavaScript rule is undefined');
						}
					});
				} else {
					throw new Error('Module is not an array');
				}
			});
		});

		test('uses style loader when in dev mode', () => {
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				getConfigFromProjectAndServer(projectConfig, serverConfig),
				'/foo/bar',
				true
			);
			const modules = cwc.getModule();
			if (Array.isArray(modules.rules)) {
				const styleRule = modules.rules.find(rule => {
					const { test } = rule;

					return (
						test !== undefined &&
						(test.toString() === '/\\.css$/' ||
							test.toString() === '/\\.(sa|sc|c)ss$/')
					);
				}) as { use: string[] };
				if (styleRule !== undefined) {
					expect(styleRule.use[0]).toBe('style-loader');
				} else {
					throw new Error('No style rule found');
				}
			} else {
				throw new Error('Module.rules is not an array');
			}
		});

		test('uses miniCssExtractPlugin when in production mode', () => {
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				getConfigFromProjectAndServer(projectConfig, serverConfig),
				'/foo/bar',
				false
			);
			const modules = cwc.getModule();
			if (Array.isArray(modules.rules)) {
				const styleRule = modules.rules.find(rule => {
					const { test } = rule;

					return (
						test !== undefined &&
						(test.toString() === '/\\.css$/' ||
							test.toString() === '/\\.(sa|sc|c)ss$/')
					);
				}) as { use: { loader: string }[] };
				if (styleRule !== undefined) {
					expect(styleRule.use[0].loader).toBe(
						miniCssExtractPlugin.loader
					);
				} else {
					throw new Error('No style rule found');
				}
			} else {
				throw new Error('Module.rules is not an array');
			}
		});

		test('has css specific file-loader option in prod mode', () => {
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				getConfigFromProjectAndServer(projectConfig, serverConfig),
				'/foo/bar',
				false
			);
			const modules = cwc.getModule();
			if (Array.isArray(modules.rules)) {
				const fileRule = modules.rules.find(rule => {
					const use = rule.use as { [_: string]: any }[];
					return (
						use !== undefined &&
						use[0].loader === 'file-loader' &&
						use[0].options.publicPath === 'assets/'
					);
				});
				expect(fileRule).not.toBeUndefined();
			} else {
				throw new Error('Module.rules is not an array');
			}
		});

		test('matches snapshot', () => {
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				getConfigFromProjectAndServer(projectConfig, serverConfig),
				'/foo/bar',
				true
			);
			const modules = cwc.getModule();
			expect(modules).toMatchSnapshot();
		});
	});

	// getResolve()
	describe('getResolve', () => {
		test('has all entry extensions and aliases', () => {
			const alias: { [x: string]: string } = {
				App: 'src/app',
				Util: 'src/util',
			};
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				{
					...getConfigFromProjectAndServer(
						projectConfig,
						serverConfig
					),
					alias: alias,
				},
				'/foo/bar',
				false
			);
			const resolve = cwc.getResolve();
			expect(resolve.extensions).toContain('.js');
			expect(resolve.extensions).toContain('.jsx');
			expect(resolve.extensions).toContain('.ts');
			expect(resolve.extensions).toContain('.tsx');
			expect(resolve.alias).toEqual(alias);
		});
	});

	// getOptimization()
	describe('getOptimization', () => {
		test('returns just runtimeChunk when not in use', () => {
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				{
					...getConfigFromProjectAndServer(
						projectConfig,
						serverConfig
					),
					optimizeSplitChunks: false,
				},
				'/foo/bar',
				false
			);
			expect(cwc.getOptimization()).toEqual({ runtimeChunk: 'single' });
		});

		test('returns object when in use', () => {
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				{
					...getConfigFromProjectAndServer(
						projectConfig,
						serverConfig
					),
					optimizeSplitChunks: true,
				},
				'/foo/bar',
				false
			);
			expect(cwc.getOptimization()).not.toBeUndefined();
			expect(
				(cwc.getOptimization() as webpack.Options.Optimization)
					.runtimeChunk
			).toBe('single');
			expect(cwc.getOptimization()).toMatchSnapshot();
		});
	});

	// getCommmon
	describe('getCommon', () => {
		test('has externals', () => {
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				{
					...getConfigFromProjectAndServer(
						projectConfig,
						serverConfig
					),
				},
				'/foo/bar',
				true
			);
			expect(cwc.getCommon().externals).toBe(projectConfig.externals);
		});
		test('sends proper stuff depending on isDev', () => {
			const cwc = new WebpackConfigHelper(
				projectConfig.files[0],
				{
					...getConfigFromProjectAndServer(
						projectConfig,
						serverConfig
					),
				},
				'/foo/bar',
				false
			);
			expect(cwc.getCommon().devtool).toBe('source-map');
			expect(cwc.getCommon()).toMatchSnapshot();
			const cwcDev = new WebpackConfigHelper(
				projectConfig.files[0],
				{
					...getConfigFromProjectAndServer(
						projectConfig,
						serverConfig
					),
				},
				'/foo/bar',
				true
			);
			expect(cwcDev.getCommon().devtool).toBe('inline-source-map');
			expect(cwcDev.getCommon()).toMatchSnapshot();
		});
	});
});
