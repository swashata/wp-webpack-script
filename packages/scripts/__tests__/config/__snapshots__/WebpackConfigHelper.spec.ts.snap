// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateWebPackConfig getCommon sends proper stuff depending on isDev 1`] = `
Object {
  "context": "/foo/bar",
  "devtool": "source-map",
  "externals": Object {
    "jquery": "jQuery",
  },
  "mode": "production",
  "name": "config1",
  "target": "web",
  "watch": false,
}
`;

exports[`CreateWebPackConfig getCommon sends proper stuff depending on isDev 2`] = `
Object {
  "context": "/foo/bar",
  "devtool": "inline-source-map",
  "externals": Object {
    "jquery": "jQuery",
  },
  "mode": "development",
  "name": "config1",
  "target": "web",
  "watch": true,
}
`;

exports[`CreateWebPackConfig getModule matches snapshot 1`] = `
Object {
  "rules": Array [
    Object {
      "exclude": /\\(node_modules\\|bower_components\\)/,
      "test": /\\\\\\.m\\?jsx\\?\\$/,
      "use": Array [
        Object {
          "loader": "babel-loader",
          "options": Object {
            "babelrc": false,
            "configFile": false,
            "presets": Array [
              Array [
                "@wpackio/base",
                Object {
                  "hasReact": true,
                  "presetEnv": Object {
                    "targets": Object {
                      "chrome": "69",
                      "edge": "17",
                      "firefox": "62",
                    },
                  },
                },
              ],
            ],
          },
        },
      ],
    },
    Object {
      "exclude": /\\(node_modules\\)/,
      "test": /\\\\\\.tsx\\?\\$/,
      "use": Array [
        Object {
          "loader": "babel-loader",
          "options": Object {
            "babelrc": false,
            "configFile": false,
            "presets": Array [
              Array [
                "@wpackio/base",
                Object {
                  "hasReact": true,
                  "presetEnv": Object {
                    "targets": Object {
                      "chrome": "69",
                      "edge": "17",
                      "firefox": "62",
                    },
                  },
                },
              ],
              Array [
                "@babel/preset-typescript",
              ],
            ],
          },
        },
      ],
    },
    Object {
      "test": /\\\\\\.\\(sa\\|sc\\|c\\)ss\\$/,
      "use": Array [
        "style-loader",
        Object {
          "loader": "css-loader",
          "options": Object {
            "importLoaders": 1,
            "sourceMap": true,
          },
        },
        "postcss-loader",
        Object {
          "loader": "sass-loader",
          "options": Object {
            "sourceMap": true,
          },
        },
      ],
    },
    Object {
      "issuer": [Function],
      "test": /\\\\\\.\\(woff\\|woff2\\|eot\\|ttf\\|otf\\|png\\|jpg\\|gif\\)\\(\\\\\\?v=\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\)\\?\\$/,
      "use": Array [
        Object {
          "loader": "file-loader",
          "options": Object {
            "name": "[name]-[hash:8].[ext]",
            "outputPath": "config1/assets/",
          },
        },
      ],
    },
    Object {
      "issuer": [Function],
      "test": /\\\\\\.\\(woff\\|woff2\\|eot\\|ttf\\|otf\\|png\\|jpg\\|gif\\)\\(\\\\\\?v=\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\)\\?\\$/,
      "use": Array [
        Object {
          "loader": "file-loader",
          "options": Object {
            "name": "[name]-[hash:8].[ext]",
            "outputPath": "config1/assets/",
            "publicPath": undefined,
          },
        },
      ],
    },
  ],
}
`;

exports[`CreateWebPackConfig getOptimization returns object when in use 1`] = `
Object {
  "runtimeChunk": "single",
  "splitChunks": Object {
    "cacheGroups": Object {
      "default": Object {
        "minChunks": 2,
        "priority": -20,
        "reuseExistingChunk": true,
      },
      "vendors": Object {
        "priority": -10,
        "test": /\\[\\\\\\\\\\\\/\\]node_modules\\[\\\\\\\\\\\\/\\]/,
      },
    },
    "chunks": "all",
    "maxAsyncRequests": 5,
    "maxInitialRequests": 3,
    "minChunks": 1,
    "minSize": 30000,
    "name": true,
  },
}
`;

exports[`CreateWebPackConfig getPlugins has proper plugins for build mode 1`] = `
Array [
  DefinePlugin {
    "definitions": Object {
      "__WPACKIO__": Object {
        "appName": "\\"wpackio\\"",
        "outputPath": "\\"dist\\"",
      },
      "process.env.BABEL_ENV": "\\"production\\"",
      "process.env.NODE_ENV": "\\"production\\"",
    },
  },
  CleanWebpackPlugin {
    "options": Object {
      "allowExternal": false,
      "dry": false,
      "root": "/foo/bar",
      "verbose": false,
    },
    "paths": Array [
      "/foo/bar/dist/config1",
    ],
  },
  MiniCssExtractPlugin {
    "options": Object {
      "chunkFilename": "config1/[name].css",
      "filename": "config1/[name].css",
    },
  },
  Object {},
  BannerPlugin {
    "banner": [Function],
    "options": Object {
      "banner": "
WordPress WebPack Bundler

@author Swashata Ghosh
@version 1.0.0
@link https://wpack.io
@license GPL-3.0

Copyright (c) 2018 Swashata Ghosh

This software is released under the GPL-3.0 License
https://opensource.org/licenses/GPL-3.0

Compiled with the help of https://wpack.io
A zero setup Webpack Bundler Script for WordPress",
      "entryOnly": false,
      "include": /\\\\\\.\\(\\(t\\|j\\)sx\\?\\|s\\?\\(c\\|a\\)ss\\)\\$/,
      "raw": false,
    },
  },
]
`;

exports[`CreateWebPackConfig getPlugins has proper plugins for dev mode 1`] = `
Array [
  TimeFixPlugin {
    "watchOffset": 11000,
  },
  DefinePlugin {
    "definitions": Object {
      "__WPACKIO__": Object {
        "appName": "\\"wpackio\\"",
        "outputPath": "\\"dist\\"",
      },
      "process.env.BABEL_ENV": "\\"development\\"",
      "process.env.NODE_ENV": "\\"development\\"",
    },
  },
  CleanWebpackPlugin {
    "options": Object {
      "allowExternal": false,
      "dry": false,
      "root": "/foo/bar",
      "verbose": false,
    },
    "paths": Array [
      "/foo/bar/dist/config1",
    ],
  },
  MiniCssExtractPlugin {
    "options": Object {
      "chunkFilename": "config1/[name].css",
      "filename": "config1/[name].css",
    },
  },
  Object {},
  HotModuleReplacementPlugin {
    "fullBuildTimeout": 200,
    "multiStep": undefined,
    "options": Object {},
    "requestTimeout": 10000,
  },
  WatchMissingNodeModulesPlugin {
    "nodeModulesPath": "/foo/bar/node_modules",
  },
]
`;
